*** orange2.0/OrangeWidgets/OWGUI.py	2010-07-29 10:21:59.000000000 +0200
--- orange2.0_PatchedOK/OrangeWidgets/OWGUI.py	2010-07-29 11:24:30.000000000 +0200
***************
*** 49,55 ****
      else:
          b = QWidget(widget)
          if margin == -1: margin = 0
!     if addToLayout and widget.layout():
          widget.layout().addWidget(b)
  
      if isinstance(orientation, QLayout):
--- 49,55 ----
      else:
          b = QWidget(widget)
          if margin == -1: margin = 0
!     if addToLayout and widget.layout() != None:
          widget.layout().addWidget(b)
  
      if isinstance(orientation, QLayout):
***************
*** 91,97 ****
          lbl = QLabel(label, widget)
          if labelWidth:
              lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
!         if widget.layout() and addToLayout: widget.layout().addWidget(lbl)
      else:
          lbl = None
  
--- 91,97 ----
          lbl = QLabel(label, widget)
          if labelWidth:
              lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
!         if widget.layout() != None and addToLayout: widget.layout().addWidget(lbl)
      else:
          lbl = None
  
***************
*** 103,109 ****
  
  def label(widget, master, label, labelWidth = None):
      lbl = QLabel("", widget)
!     if widget.layout(): widget.layout().addWidget(lbl)
  
      reprint = CallFrontLabel(lbl, label, master)
      for mo in __re_frmt.finditer(label):
--- 103,109 ----
  
  def label(widget, master, label, labelWidth = None):
      lbl = QLabel("", widget)
!     if widget.layout() != None: widget.layout().addWidget(lbl)
  
      reprint = CallFrontLabel(lbl, label, master)
      for mo in __re_frmt.finditer(label):
***************
*** 183,189 ****
  
      wa = bi.control = SpinBoxWFocusOut(min, max, step, bi)
      wa.setAlignment(alignment)
!     if bi.layout(): bi.layout().addWidget(wa)
      # must be defined because of the setText below
      if controlWidth:
          wa.setFixedWidth(controlWidth)
--- 183,189 ----
  
      wa = bi.control = SpinBoxWFocusOut(min, max, step, bi)
      wa.setAlignment(alignment)
!     if bi.layout() != None: bi.layout().addWidget(wa)
      # must be defined because of the setText below
      if controlWidth:
          wa.setFixedWidth(controlWidth)
***************
*** 277,283 ****
  
      wa = bi.control = DoubleSpinBoxWFocusOut(min, max, step, bi)
      wa.setAlignment(alignment)
!     if addToLayout and bi.layout(): bi.layout().addWidget(wa)
      # must be defined because of the setText below
      if controlWidth:
          wa.setFixedWidth(controlWidth)
--- 277,283 ----
  
      wa = bi.control = DoubleSpinBoxWFocusOut(min, max, step, bi)
      wa.setAlignment(alignment)
!     if addToLayout and bi.layout() != None: bi.layout().addWidget(wa)
      # must be defined because of the setText below
      if controlWidth:
          wa.setFixedWidth(controlWidth)
***************
*** 319,325 ****
  def checkBox(widget, master, value, label, box=None, tooltip=None, callback=None, getwidget=None, id=None, disabled=0, labelWidth=None, disables = [], addToLayout = 1, debuggingEnabled = 1):
      b = widgetBox(widget, box, orientation=None)
      wa = QCheckBox(label, b)
!     if addToLayout and b.layout(): b.layout().addWidget(wa)
  
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
--- 319,325 ----
  def checkBox(widget, master, value, label, box=None, tooltip=None, callback=None, getwidget=None, id=None, disabled=0, labelWidth=None, disables = [], addToLayout = 1, debuggingEnabled = 1):
      b = widgetBox(widget, box, orientation=None)
      wa = QCheckBox(label, b)
!     if addToLayout and b.layout() != None: b.layout().addWidget(wa)
  
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
***************
*** 344,350 ****
      button = QPushButton(parent)
      button.setFixedSize(height, height)
      button.setIcon(getEnterIcon())
!     if parent.layout(): parent.layout().addWidget(button)
      if not placeholder:
          return button
  
--- 344,350 ----
      button = QPushButton(parent)
      button.setFixedSize(height, height)
      button.setIcon(getEnterIcon())
!     if parent.layout() != None: parent.layout().addWidget(button)
      if not placeholder:
          return button
  
***************
*** 419,425 ****
          wa = QLineEdit(b)
          wa.enterButton = None
  
!     if b and b.layout(): b.layout().addWidget(wa)
      if value:
          wa.setText(unicode(getdeepattr(master, value)))
  
--- 419,425 ----
          wa = QLineEdit(b)
          wa.enterButton = None
  
!     if b and b.layout() != None: b.layout().addWidget(wa)
      if value:
          wa.setText(unicode(getdeepattr(master, value)))
  
***************
*** 440,446 ****
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, debuggingEnabled = 1, width = None, height = None, toggleButton = False, value = "", addToLayout = 1):
      btn = QPushButton(label, widget)
!     if addToLayout and widget.layout():
          widget.layout().addWidget(btn)
  
      if width:
--- 440,446 ----
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, debuggingEnabled = 1, width = None, height = None, toggleButton = False, value = "", addToLayout = 1):
      btn = QPushButton(label, widget)
!     if addToLayout and widget.layout() != None:
          widget.layout().addWidget(btn)
  
      if width:
***************
*** 467,473 ****
  
  def toolButton(widget, master, callback = None, width = None, height = None, tooltip = None, addToLayout = 1, debuggingEnabled = 1):
      btn = QToolButton(widget)
!     if addToLayout and widget.layout(): widget.layout().addWidget(btn)
      if width != None: btn.setFixedWidth(width)
      if height!= None: btn.setFixedHeight(height)
      if tooltip != None: btn.setToolTip(tooltip)
--- 467,473 ----
  
  def toolButton(widget, master, callback = None, width = None, height = None, tooltip = None, addToLayout = 1, debuggingEnabled = 1):
      btn = QToolButton(widget)
!     if addToLayout and widget.layout() != None: widget.layout().addWidget(btn)
      if width != None: btn.setFixedWidth(width)
      if height!= None: btn.setFixedHeight(height)
      if tooltip != None: btn.setToolTip(tooltip)
***************
*** 480,486 ****
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout(): widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
--- 480,486 ----
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout() != None: widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
***************
*** 518,524 ****
      lb = OrangeListBox(master, value, enableDragDrop, dragDropCallback, dataValidityCallback, sizeHint, bg)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout(): bg.layout().addWidget(lb)
  
      if value != None:
          clist = getdeepattr(master, value)
--- 518,524 ----
      lb = OrangeListBox(master, value, enableDragDrop, dragDropCallback, dataValidityCallback, sizeHint, bg)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout()  != None: bg.layout().addWidget(lb)
  
      if value != None:
          clist = getdeepattr(master, value)
***************
*** 584,590 ****
          w = QRadioButton(unicode(i))
          w.setIcon(QIcon(label))
      #w.ogValue = value
!     if addToLayout and dest.layout(): dest.layout().addWidget(w)
      if not hasattr(bg, "group"):
          bg.group = QButtonGroup(bg)
      bg.group.addButton(w)
--- 584,590 ----
          w = QRadioButton(unicode(i))
          w.setIcon(QIcon(label))
      #w.ogValue = value
!     if addToLayout and dest.layout() != None: dest.layout().addWidget(w)
      if not hasattr(bg, "group"):
          bg.group = QButtonGroup(bg)
      bg.group.addButton(w)
***************
*** 610,616 ****
  #    else:
  #        w = QRadioButton("X")
  #        w.setPixmap(label)
! #    if bg.layout(): bg.layout().addWidget(w)
  #
  #    w.setChecked(getdeepattr(master, value))
  #    if tooltip:
--- 610,616 ----
  #    else:
  #        w = QRadioButton("X")
  #        w.setPixmap(label)
! #    if bg.layout() != None: bg.layout().addWidget(w)
  #
  #    w.setChecked(getdeepattr(master, value))
  #    if tooltip:
***************
*** 645,651 ****
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout(): sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
--- 645,651 ----
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout() != None: sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
***************
*** 653,659 ****
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout(): sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
--- 653,659 ----
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout() != None: sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
***************
*** 881,887 ****
              QPushButton.__init__(self, QIcon(name), text, widget)
          else:
              QPushButton.__init__(self, text, widget)
!         if widget.layout():
              widget.layout().addWidget(self)
          if tooltip != None:
              self.setToolTip(tooltip)
--- 881,887 ----
              QPushButton.__init__(self, QIcon(name), text, widget)
          else:
              QPushButton.__init__(self, text, widget)
!         if widget.layout() != None:
              widget.layout().addWidget(self)
          if tooltip != None:
              self.setToolTip(tooltip)
***************
*** 934,940 ****
              else:
                  name = os.path.join(iconDir, "arrow_down.png")
              self.setPixmap(QPixmap(name))
!         if widget.layout():
              widget.layout().addWidget(self)
          if tooltip != None:
              self.setToolTip(tooltip)
--- 934,940 ----
              else:
                  name = os.path.join(iconDir, "arrow_down.png")
              self.setPixmap(QPixmap(name))
!         if widget.layout() != None:
              widget.layout().addWidget(self)
          if tooltip != None:
              self.setToolTip(tooltip)
***************
*** 1065,1071 ****
      if indent:
          hb = widgetBox(hb, orientation = "horizontal")
          hb.layout().addSpacing(indent)
!     if hb.layout() and addToLayout: hb.layout().addWidget(combo)
  
      if items:
          combo.addItems([unicode(i) for i in items])
--- 1065,1071 ----
      if indent:
          hb = widgetBox(hb, orientation = "horizontal")
          hb.layout().addSpacing(indent)
!     if hb.layout() != None and addToLayout: hb.layout().addWidget(combo)
  
      if items:
          combo.addItems([unicode(i) for i in items])
***************
*** 1108,1114 ****
          if orientation == 'vertical': self.setLayout(QVBoxLayout())
          else:                         self.setLayout(QHBoxLayout())
  
!         if widget.layout():
              widget.layout().addWidget(self)
          if type(box) in (str, unicode): # if you pass 1 for box, there will be a box, but no text
              self.setTitle(" " + box.strip() + " ")
--- 1108,1114 ----
          if orientation == 'vertical': self.setLayout(QVBoxLayout())
          else:                         self.setLayout(QHBoxLayout())
  
!         if widget.layout() != None:
              widget.layout().addWidget(self)
          if type(box) in (str, unicode): # if you pass 1 for box, there will be a box, but no text
              self.setTitle(" " + box.strip() + " ")
***************
*** 1150,1156 ****
  class widgetHider(QWidget):
      def __init__(self, widget, master, value, size = (19,19), widgets = [], tooltip = None):
          QWidget.__init__(self, widget)
!         if widget.layout():
              widget.layout().addWidget(self)
          self.value = value
          self.master = master
--- 1150,1156 ----
  class widgetHider(QWidget):
      def __init__(self, widget, master, value, size = (19,19), widgets = [], tooltip = None):
          QWidget.__init__(self, widget)
!         if widget.layout() != None:
              widget.layout().addWidget(self)
          self.value = value
          self.master = master
***************
*** 1817,1823 ****
  
  def tabWidget(widget):
      w = QTabWidget(widget)
!     if widget.layout():
          widget.layout().addWidget(w)
      return w
  
--- 1817,1823 ----
  
  def tabWidget(widget):
      w = QTabWidget(widget)
!     if widget.layout() != None:
          widget.layout().addWidget(w)
      return w
  
***************
*** 1837,1843 ****
  
  def table(widget, rows = 0, columns = 0, selectionMode = -1, addToLayout = 1):
      w = QTableWidget(rows, columns, widget)
!     if widget and addToLayout and widget.layout():
          widget.layout().addWidget(w)
      if selectionMode != -1:
          w.setSelectionMode(selectionMode)   
--- 1837,1843 ----
  
  def table(widget, rows = 0, columns = 0, selectionMode = -1, addToLayout = 1):
      w = QTableWidget(rows, columns, widget)
!     if widget and addToLayout and widget.layout() != None:
          widget.layout().addWidget(w)
      if selectionMode != -1:
          w.setSelectionMode(selectionMode)   
*** orange2.0/OrangeCanvas/orngGui.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeCanvas/orngGui.py	2010-07-29 11:27:01.000000000 +0200
***************
*** 3,9 ****
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout(): widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
--- 3,9 ----
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout() != None: widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
***************
*** 15,21 ****
              b.setTitle(" "+box.strip()+" ")
      else:
          b = QWidget(widget)
!     if widget.layout(): widget.layout().addWidget(b)
  
      if orientation == 'horizontal' or not orientation:
          b.setLayout(QHBoxLayout())
--- 15,21 ----
              b.setTitle(" "+box.strip()+" ")
      else:
          b = QWidget(widget)
!     if widget.layout() != None: widget.layout().addWidget(b)
  
      if orientation == 'horizontal' or not orientation:
          b.setLayout(QHBoxLayout())
***************
*** 32,38 ****
  
  def widgetLabel(widget, label, labelWidth=None):
      lbl = QLabel(label, widget)
!     if widget.layout(): widget.layout().addWidget(lbl)
      if labelWidth:
          lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
      return lbl
--- 32,38 ----
  
  def widgetLabel(widget, label, labelWidth=None):
      lbl = QLabel(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(lbl)
      if labelWidth:
          lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
      return lbl
***************
*** 42,55 ****
          widget = widgetBox(widget, box, orientation=None)
      if indent:
          hbox = widgetBox(widget, orientation = "horizontal")
!         if widget.layout(): widget.layout().addWidget(hbox)
          sep = QWidget(hbox)
          sep.setFixedSize(indent, 5)
!         if hbox.layout(): hbox.layout().addWidget(sep)
          widget = hbox
  
      wa = QCheckBox(label, widget)
!     if widget.layout(): widget.layout().addWidget(wa)
       
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
--- 42,55 ----
          widget = widgetBox(widget, box, orientation=None)
      if indent:
          hbox = widgetBox(widget, orientation = "horizontal")
!         if widget.layout() != None: widget.layout().addWidget(hbox)
          sep = QWidget(hbox)
          sep.setFixedSize(indent, 5)
!         if hbox.layout() != None: hbox.layout().addWidget(sep)
          widget = hbox
  
      wa = QCheckBox(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(wa)
       
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
***************
*** 64,76 ****
      b = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(b, label, labelWidth)
!         if b.layout(): b.layout().addWidget(l)
  
      if args.has_key("baseClass"):
          wa = args["baseClass"](b)
      else:
          wa = QLineEdit(b)
!     if b.layout(): b.layout().addWidget(wa)
  
      if controlWidth:
          wa.setFixedWidth(controlWidth)
--- 64,76 ----
      b = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(b, label, labelWidth)
!         if b.layout() != None: b.layout().addWidget(l)
  
      if args.has_key("baseClass"):
          wa = args["baseClass"](b)
      else:
          wa = QLineEdit(b)
!     if b.layout() != None: b.layout().addWidget(wa)
  
      if controlWidth:
          wa.setFixedWidth(controlWidth)
***************
*** 85,91 ****
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, width = None):
      btn = QPushButton(label, widget)
!     if widget.layout(): widget.layout().addWidget(btn)
  
      if width:
          btn.setFixedWidth(width)
--- 85,91 ----
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, width = None):
      btn = QPushButton(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(btn)
  
      if width:
          btn.setFixedWidth(width)
***************
*** 102,108 ****
      lb = QListWidget(master)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout(): bg.layout().addWidget(lb)
  
      if tooltip:
          lb.setToolTip(tooltip)
--- 102,108 ----
      lb = QListWidget(master)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout() != None: bg.layout().addWidget(lb)
  
      if tooltip:
          lb.setToolTip(tooltip)
***************
*** 117,125 ****
      hb = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(hb, label, labelWidth)
!         if hb.layout(): hb.layout().addWidget(l)
      combo = QComboBox(hb)
!     if hb.layout(): hb.layout().addWidget(combo)
  
      if tooltip:
          combo.setToolTip(tooltip)
--- 117,125 ----
      hb = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(hb, label, labelWidth)
!         if hb.layout() != None: hb.layout().addWidget(l)
      combo = QComboBox(hb)
!     if hb.layout() != None: hb.layout().addWidget(combo)
  
      if tooltip:
          combo.setToolTip(tooltip)
***************
*** 133,141 ****
  def comboBoxWithCaption(widget, master, value, label, box=None, items=None, tooltip=None, callback = None, labelWidth = None):
      hb = widgetBox(widget, box = box, orientation="horizontal")
      l = widgetLabel(hb, label, labelWidth)
!     if hb.layout(): hb.layout().addWidget(l)
      combo = comboBox(hb, master, value, items = items, tooltip = tooltip, callback = callback)
!     if hb.layout(): hb.layout().addWidget(combo)
      return combo
  
  
--- 133,141 ----
  def comboBoxWithCaption(widget, master, value, label, box=None, items=None, tooltip=None, callback = None, labelWidth = None):
      hb = widgetBox(widget, box = box, orientation="horizontal")
      l = widgetLabel(hb, label, labelWidth)
!     if hb.layout() != None: hb.layout().addWidget(l)
      combo = comboBox(hb, master, value, items = items, tooltip = tooltip, callback = callback)
!     if hb.layout() != None: hb.layout().addWidget(combo)
      return combo
  
  
***************
*** 162,168 ****
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout(): sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
--- 162,168 ----
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout() != None: sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
***************
*** 170,176 ****
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout(): sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
--- 170,176 ----
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout() != None: sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
*** orange2.0/OrangeWidgets/ColorPalette.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeWidgets/ColorPalette.py	2010-07-29 11:53:48.000000000 +0200
***************
*** 283,289 ****
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          self.setMaximumWidth(135)
!         if parent and parent.layout():
              parent.layout().addWidget(self)
  
      def setPalette1(self, palette):
--- 283,289 ----
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          self.setMaximumWidth(135)
!         if parent and parent.layout() != None:
              parent.layout().addWidget(self)
  
      def setPalette1(self, palette):
***************
*** 305,311 ****
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout():
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
--- 305,311 ----
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout() != None:
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
*** orange2.0/OrangeWidgets/OWColorPalette.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeWidgets/OWColorPalette.py	2010-07-29 11:54:36.000000000 +0200
***************
*** 648,654 ****
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          
!         if parent and parent.layout():
              parent.layout().addWidget(self)
      
      def resizeEvent(self, ev):
--- 648,654 ----
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          
!         if parent and parent.layout() != None:
              parent.layout().addWidget(self)
      
      def resizeEvent(self, ev):
***************
*** 739,745 ****
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout():
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
--- 739,745 ----
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout() != None:
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
*** orange2.0/OrangeWidgets/OWToolbars.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeWidgets/OWToolbars.py	2010-07-29 12:00:53.000000000 +0200
***************
*** 26,32 ****
  def createButton(parent, text, action = None, icon = None, toggle = 0):
      btn = QToolButton(parent)
      btn.setMinimumSize(30,30)
!     if parent.layout():
          parent.layout().addWidget(btn)
      btn.setCheckable(toggle)
      if action:
--- 26,32 ----
  def createButton(parent, text, action = None, icon = None, toggle = 0):
      btn = QToolButton(parent)
      btn.setMinimumSize(30,30)
!     if parent.layout() != None:
          parent.layout().addWidget(btn)
      btn.setCheckable(toggle)
      if action:
***************
*** 64,70 ****
          self.setLayout(QHBoxLayout())
          self.layout().setMargin(6)
          self.layout().setSpacing(4)
!         if parent.layout():
              parent.layout().addWidget(self)
  
          self.graph = graph # save graph. used to send signals
--- 64,70 ----
          self.setLayout(QHBoxLayout())
          self.layout().setMargin(6)
          self.layout().setSpacing(4)
!         if parent.layout() != None:
              parent.layout().addWidget(self)
  
          self.graph = graph # save graph. used to send signals
***************
*** 140,146 ****
          self.layout().setSpacing(0)
          self.layout().setContentsMargins(0,25,0,0)
          
!         if parent.layout():
              parent.layout().addWidget(self)
  
          self.navigate = OWGUI.widgetBox(self, 0, orientation = "vertical", margin=2)
--- 140,146 ----
          self.layout().setSpacing(0)
          self.layout().setContentsMargins(0,25,0,0)
          
!         if parent.layout() != None:
              parent.layout().addWidget(self)
  
          self.navigate = OWGUI.widgetBox(self, 0, orientation = "vertical", margin=2)
