*** orange2.0/OrangeCanvas/orngGui.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeCanvas/orngGui.py	2010-07-29 11:27:01.000000000 +0200
***************
*** 3,9 ****
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout(): widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
--- 3,9 ----
  
  def separator(widget, width=8, height=8):
      sep = QWidget(widget)
!     if widget.layout() != None: widget.layout().addWidget(sep)
      sep.setFixedSize(width, height)
      return sep
  
***************
*** 15,21 ****
              b.setTitle(" "+box.strip()+" ")
      else:
          b = QWidget(widget)
!     if widget.layout(): widget.layout().addWidget(b)
  
      if orientation == 'horizontal' or not orientation:
          b.setLayout(QHBoxLayout())
--- 15,21 ----
              b.setTitle(" "+box.strip()+" ")
      else:
          b = QWidget(widget)
!     if widget.layout() != None: widget.layout().addWidget(b)
  
      if orientation == 'horizontal' or not orientation:
          b.setLayout(QHBoxLayout())
***************
*** 32,38 ****
  
  def widgetLabel(widget, label, labelWidth=None):
      lbl = QLabel(label, widget)
!     if widget.layout(): widget.layout().addWidget(lbl)
      if labelWidth:
          lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
      return lbl
--- 32,38 ----
  
  def widgetLabel(widget, label, labelWidth=None):
      lbl = QLabel(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(lbl)
      if labelWidth:
          lbl.setFixedSize(labelWidth, lbl.sizeHint().height())
      return lbl
***************
*** 42,55 ****
          widget = widgetBox(widget, box, orientation=None)
      if indent:
          hbox = widgetBox(widget, orientation = "horizontal")
!         if widget.layout(): widget.layout().addWidget(hbox)
          sep = QWidget(hbox)
          sep.setFixedSize(indent, 5)
!         if hbox.layout(): hbox.layout().addWidget(sep)
          widget = hbox
  
      wa = QCheckBox(label, widget)
!     if widget.layout(): widget.layout().addWidget(wa)
       
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
--- 42,55 ----
          widget = widgetBox(widget, box, orientation=None)
      if indent:
          hbox = widgetBox(widget, orientation = "horizontal")
!         if widget.layout() != None: widget.layout().addWidget(hbox)
          sep = QWidget(hbox)
          sep.setFixedSize(indent, 5)
!         if hbox.layout() != None: hbox.layout().addWidget(sep)
          widget = hbox
  
      wa = QCheckBox(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(wa)
       
      if labelWidth:
          wa.setFixedSize(labelWidth, wa.sizeHint().height())
***************
*** 64,76 ****
      b = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(b, label, labelWidth)
!         if b.layout(): b.layout().addWidget(l)
  
      if args.has_key("baseClass"):
          wa = args["baseClass"](b)
      else:
          wa = QLineEdit(b)
!     if b.layout(): b.layout().addWidget(wa)
  
      if controlWidth:
          wa.setFixedWidth(controlWidth)
--- 64,76 ----
      b = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(b, label, labelWidth)
!         if b.layout() != None: b.layout().addWidget(l)
  
      if args.has_key("baseClass"):
          wa = args["baseClass"](b)
      else:
          wa = QLineEdit(b)
!     if b.layout() != None: b.layout().addWidget(wa)
  
      if controlWidth:
          wa.setFixedWidth(controlWidth)
***************
*** 85,91 ****
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, width = None):
      btn = QPushButton(label, widget)
!     if widget.layout(): widget.layout().addWidget(btn)
  
      if width:
          btn.setFixedWidth(width)
--- 85,91 ----
  
  def button(widget, master, label, callback = None, disabled=0, tooltip=None, width = None):
      btn = QPushButton(label, widget)
!     if widget.layout() != None: widget.layout().addWidget(btn)
  
      if width:
          btn.setFixedWidth(width)
***************
*** 102,108 ****
      lb = QListWidget(master)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout(): bg.layout().addWidget(lb)
  
      if tooltip:
          lb.setToolTip(tooltip)
--- 102,108 ----
      lb = QListWidget(master)
      lb.box = bg
      lb.setSelectionMode(selectionMode)
!     if bg.layout() != None: bg.layout().addWidget(lb)
  
      if tooltip:
          lb.setToolTip(tooltip)
***************
*** 117,125 ****
      hb = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(hb, label, labelWidth)
!         if hb.layout(): hb.layout().addWidget(l)
      combo = QComboBox(hb)
!     if hb.layout(): hb.layout().addWidget(combo)
  
      if tooltip:
          combo.setToolTip(tooltip)
--- 117,125 ----
      hb = widgetBox(widget, box, orientation)
      if label:
          l = widgetLabel(hb, label, labelWidth)
!         if hb.layout() != None: hb.layout().addWidget(l)
      combo = QComboBox(hb)
!     if hb.layout() != None: hb.layout().addWidget(combo)
  
      if tooltip:
          combo.setToolTip(tooltip)
***************
*** 133,141 ****
  def comboBoxWithCaption(widget, master, value, label, box=None, items=None, tooltip=None, callback = None, labelWidth = None):
      hb = widgetBox(widget, box = box, orientation="horizontal")
      l = widgetLabel(hb, label, labelWidth)
!     if hb.layout(): hb.layout().addWidget(l)
      combo = comboBox(hb, master, value, items = items, tooltip = tooltip, callback = callback)
!     if hb.layout(): hb.layout().addWidget(combo)
      return combo
  
  
--- 133,141 ----
  def comboBoxWithCaption(widget, master, value, label, box=None, items=None, tooltip=None, callback = None, labelWidth = None):
      hb = widgetBox(widget, box = box, orientation="horizontal")
      l = widgetLabel(hb, label, labelWidth)
!     if hb.layout() != None: hb.layout().addWidget(l)
      combo = comboBox(hb, master, value, items = items, tooltip = tooltip, callback = callback)
!     if hb.layout() != None: hb.layout().addWidget(combo)
      return combo
  
  
***************
*** 162,168 ****
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout(): sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
--- 162,168 ----
      if width != None:
          slider.setFixedWidth(width)
  
!     if sliderBox.layout() != None: sliderBox.layout().addWidget(slider)
  
      if ticks:
          slider.setTickPosition(QSlider.TicksBelow)
***************
*** 170,176 ****
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout(): sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
--- 170,176 ----
  
      if createLabel:
          label = QLabel(sliderBox)
!         if sliderBox.layout() != None: sliderBox.layout().addWidget(label)
          label.setText(labelFormat % minValue)
          width1 = label.sizeHint().width()
          label.setText(labelFormat % maxValue)
*** orange2.0/OrangeWidgets/ColorPalette.py	2010-07-29 10:22:00.000000000 +0200
--- orange2.0_PatchedOK/OrangeWidgets/ColorPalette.py	2010-07-29 11:53:48.000000000 +0200
***************
*** 283,289 ****
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          self.setMaximumWidth(135)
!         if parent and parent.layout():
              parent.layout().addWidget(self)
  
      def setPalette1(self, palette):
--- 283,289 ----
          self.setFixedHeight(colorButtonSize)
          self.setMinimumWidth(colorButtonSize)
          self.setMaximumWidth(135)
!         if parent and parent.layout() != None:
              parent.layout().addWidget(self)
  
      def setPalette1(self, palette):
***************
*** 305,311 ****
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout():
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
--- 305,311 ----
          self.parent = parent
          self.master = master
  
!         if self.parent and self.parent.layout() != None:
              self.parent.layout().addWidget(self)
  
          self.setLayout(QHBoxLayout())
